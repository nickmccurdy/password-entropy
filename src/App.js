import CustomOptions from './CustomOptions'
import DicewareOptions from './DicewareOptions'
import GenericOptions from './GenericOptions'
import React, { PureComponent } from 'react'
import { Tab, Tabs, TabList, TabPanel } from 'react-tabs'
import 'react-tabs/style/react-tabs.css'

export default class App extends PureComponent {
  state = {
    length: 6
  }

  handleLengthChange = event => this.setState({ length: parseInt(event.target.value, 10) })
  handlePossibleItemsChange = possibleItems => this.setState({ possibleItems })

  render () {
    const possiblePasswords = (this.state.possibleItems || 1) ** this.state.length
    const approximatePrefix = possiblePasswords > Number.MAX_SAFE_INTEGER && '~ '

    return (
      <form>
        <h1>Password Entropy</h1>

        <label>
          <h2>Length</h2>
          <input value={this.state.length} onChange={this.handleLengthChange} type="number" min="1" required/>
        </label>

        <h2>Options</h2>

        <Tabs>
          <TabList>
            <Tab>Diceware</Tab>
            <Tab>Custom</Tab>
            <Tab>Generic</Tab>
          </TabList>

          <TabPanel>
            <DicewareOptions onChange={this.handlePossibleItemsChange}/>
          </TabPanel>

          <TabPanel>
            <CustomOptions onChange={this.handlePossibleItemsChange}/>
          </TabPanel>

          <TabPanel>
            <GenericOptions onChange={this.handlePossibleItemsChange}/>
          </TabPanel>
        </Tabs>

        <h2>Possible Passwords</h2>
        {approximatePrefix}{possiblePasswords.toLocaleString()} ({approximatePrefix}{Math.log2(possiblePasswords).toFixed(2)} bits of entropy)
      </form>
    )
  }
}
