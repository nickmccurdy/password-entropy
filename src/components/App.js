import {
  AppBar,
  Card,
  CardContent,
  Grid,
  Tab,
  Tabs,
  TextField,
  Toolbar,
  Typography
} from 'material-ui'
import PropTypes from 'prop-types'
import React, { Fragment } from 'react'

import Options from './Options'
import ResultsContainer from './ResultsContainer'

export default function App({
  length,
  onCustom,
  onGeneric,
  onLength,
  onOptionsKey,
  options,
  optionsKey
}) {
  return (
    <Fragment>
      <AppBar position="static">
        <Toolbar>
          <Typography variant="headline" color="inherit">
            Password Entropy
          </Typography>
        </Toolbar>

        <Tabs value={optionsKey} onChange={onOptionsKey}>
          {['Generic', 'Diceware', 'Custom'].map(name => (
            <Tab key={name} label={name} value={name} />
          ))}
        </Tabs>
      </AppBar>

      <Grid container>
        <Grid item xs={12} sm>
          <Card>
            <CardContent>
              <Typography variant="title" gutterBottom>
                Options
              </Typography>

              <TextField
                label="Number of Items"
                value={length}
                onChange={onLength}
                type="number"
                min="0"
                autoFocus
                fullWidth
              />

              <Options
                onCustom={onCustom}
                onGeneric={onGeneric}
                options={options}
                optionsKey={optionsKey}
              />
            </CardContent>
          </Card>
        </Grid>

        <Grid item xs={12} sm>
          <ResultsContainer
            length={length}
            options={options}
            optionsKey={optionsKey}
          />
        </Grid>
      </Grid>
    </Fragment>
  )
}

App.propTypes = {
  length: PropTypes.number.isRequired,
  onCustom: PropTypes.func.isRequired,
  onGeneric: PropTypes.func.isRequired,
  onLength: PropTypes.func.isRequired,
  onOptionsKey: PropTypes.func.isRequired,
  options: PropTypes.shape({
    Custom: PropTypes.number.isRequired,
    Generic: PropTypes.shape({
      'Lowercase Letters': PropTypes.bool.isRequired,
      'Uppercase Letters': PropTypes.bool.isRequired,
      Numbers: PropTypes.bool.isRequired,
      Symbols: PropTypes.bool.isRequired
    }).isRequired
  }).isRequired,
  optionsKey: PropTypes.string.isRequired
}
