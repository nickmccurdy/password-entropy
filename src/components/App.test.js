import { shallow } from 'enzyme'
import React from 'react'

import App from './App'

test('App', () => {
  const handleLength = jest.fn()
  const handleOptionsKey = jest.fn()
  const wrapper = shallow(
    <App
      length={12}
      onCustom={() => {}}
      onGeneric={() => {}}
      onLength={handleLength}
      onOptionsKey={handleOptionsKey}
      options={{
        Custom: 0,
        Generic: {
          'Lowercase Letters': true,
          'Uppercase Letters': true,
          Numbers: true,
          Symbols: true
        }
      }}
      optionsKey="Generic"
    />
  )
  wrapper.find('TextField').simulate('change')
  wrapper.find('WithStyles(Tabs)').simulate('change')
  expect(wrapper).toMatchSnapshot()
  expect(handleLength).toHaveBeenCalled()
  expect(handleOptionsKey).toHaveBeenCalled()
})
