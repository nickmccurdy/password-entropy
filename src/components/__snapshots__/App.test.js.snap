// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App 1`] = `
<React.Fragment>
  <WithStyles(AppBar)
    position="static"
  >
    <WithStyles(Toolbar)>
      <WithStyles(Typography)
        color="inherit"
        variant="headline"
      >
        Password Entropy
      </WithStyles(Typography)>
    </WithStyles(Toolbar)>
    <WithStyles(Tabs)
      onChange={[Function]}
      value="Generic"
    >
      <WithStyles(Tab)
        key="Generic"
        label="Generic"
        value="Generic"
      />
      <WithStyles(Tab)
        key="Diceware"
        label="Diceware"
        value="Diceware"
      />
      <WithStyles(Tab)
        key="Custom"
        label="Custom"
        value="Custom"
      />
    </WithStyles(Tabs)>
  </WithStyles(AppBar)>
  <WithStyles(GridWrapper)
    container={true}
  >
    <WithStyles(GridWrapper)
      item={true}
      sm={true}
      xs={12}
    >
      <Options
        length={12}
        onCustom={[Function]}
        onGeneric={[Function]}
        onLength={[Function]}
        options={
          Object {
            "Custom": 0,
            "Generic": Object {
              "Lowercase Letters": true,
              "Numbers": true,
              "Symbols": true,
              "Uppercase Letters": true,
            },
          }
        }
        optionsKey="Generic"
      />
    </WithStyles(GridWrapper)>
    <WithStyles(GridWrapper)
      item={true}
      sm={true}
      xs={12}
    >
      <ResultsContainer
        length={12}
        options={
          Object {
            "Custom": 0,
            "Generic": Object {
              "Lowercase Letters": true,
              "Numbers": true,
              "Symbols": true,
              "Uppercase Letters": true,
            },
          }
        }
        optionsKey="Generic"
      />
    </WithStyles(GridWrapper)>
  </WithStyles(GridWrapper)>
</React.Fragment>
`;
